File: clients/c-client-app/src/my-component.c
#include "../my_world.h"
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

typedef struct {
    float x;
    float y;
    float speed;
    float radius;
    char color[20];  // Assuming a maximum of 20 characters for the color string
} Particle;

Particle particles[100];  // Fixed-size array for particles
int particleCount = 0;

unsigned int tick = 0;

float heart_offset[2] = {0.0, 0.0};

void createParticles() {
    levo_portal_my_imports_size_t canvasSize;
    levo_portal_my_imports_canvas_size(&canvasSize);
    float canvas_width = canvasSize.width;

    tick++;
    
    levo_portal_my_imports_mouse_button_t leftMouseButton;
    leftMouseButton.tag = LEVO_PORTAL_MY_IMPORTS_MOUSE_BUTTON_LEFT;

    if (tick % 10 == 0 && particleCount < 100) {
        particles[particleCount].x = ((float)rand() / RAND_MAX) * canvas_width;
        particles[particleCount].y = 0.0;
        particles[particleCount].speed = 500.0 + ((float)rand() / RAND_MAX) * 13.0;
        particles[particleCount].radius = 5.0 + ((float)rand() / RAND_MAX) * 5.0;

        const char* color = "white";
        strncpy(particles[particleCount].color, color, sizeof(particles[particleCount].color) - 1);
        particles[particleCount].color[sizeof(particles[particleCount].color) - 1] = '\0';  // Ensure null-termination

        if (levo_portal_my_imports_mouse_button_pressed(&leftMouseButton)) {
            levo_portal_my_imports_position_t cursor_position;
            levo_portal_my_imports_cursor_position(&cursor_position);
            particles[particleCount].x = cursor_position.x;
            particles[particleCount].y = cursor_position.y;
        }

        particleCount++;
    }
}

void updateParticles() {
    for (int i = 0; i < particleCount; i++) {
        particles[i].y -= particles[i].speed * levo_portal_my_imports_delta_seconds();
    }
}

void killParticles() {
    levo_portal_my_imports_size_t canvasSize;
    levo_portal_my_imports_canvas_size(&canvasSize);
    float canvas_height = canvasSize.height;
    for (int i = 0; i < particleCount; i++) {
        if (particles[i].y < -canvas_height) {
            particles[i].y = 0.0;
        }
    }
}

void drawParticles() {
    my_world_string_t fill_style;
    my_world_string_set(&fill_style, "royal_purple");
    levo_portal_my_imports_fill_style(&fill_style);
    levo_portal_my_imports_size_t canvasSize;
    levo_portal_my_imports_canvas_size(&canvasSize);
    float canvas_height = canvasSize.height;
    levo_portal_my_imports_fill_rect(0.0, 0.0, canvasSize.width, canvasSize.height);

    for (int i = 0; i < particleCount; i++) {
        levo_portal_my_imports_begin_path();
        levo_portal_my_imports_arc(
            particles[i].x,
            particles[i].y,
            particles[i].radius,
            2.0 * M_PI,
            0.0
        );
        levo_portal_my_imports_close_path();
        my_world_string_t fill_style;
        my_world_string_set(&fill_style, particles[i].color);
        levo_portal_my_imports_fill_style(&fill_style);
        levo_portal_my_imports_fill();
    }
}

void drawHeart(float x_offset, float y_offset) {
    levo_portal_my_imports_begin_path();
    levo_portal_my_imports_move_to(x_offset, y_offset);
    levo_portal_my_imports_cubic_bezier_to(x_offset + 75.0, y_offset + 75.0, x_offset + 175.0, y_offset - 50.0, x_offset, y_offset - 150.0);
    levo_portal_my_imports_cubic_bezier_to(x_offset - 175.0, y_offset - 50.0, x_offset - 75.0, y_offset + 75.0, x_offset, y_offset);
    levo_portal_my_imports_close_path();

    my_world_string_t fill_style;
    my_world_string_set(&fill_style, "red");
    levo_portal_my_imports_fill_style(&fill_style);
 
    levo_portal_my_imports_fill();
    my_world_string_free(&fill_style);
}

void my_world_update() {
    createParticles();
    updateParticles();
    killParticles();
    drawParticles();

    if (tick > 100) {
        float heart_speed = 222.0;
    
        if (levo_portal_my_imports_key_pressed(LEVO_PORTAL_MY_IMPORTS_KEY_CODE_LEFT)) {
            heart_offset[0] -= heart_speed * levo_portal_my_imports_delta_seconds();
        }
    
        if (levo_portal_my_imports_key_pressed(LEVO_PORTAL_MY_IMPORTS_KEY_CODE_RIGHT)) {
            heart_offset[0] += heart_speed * levo_portal_my_imports_delta_seconds();
        }
    
        if (levo_portal_my_imports_key_pressed(LEVO_PORTAL_MY_IMPORTS_KEY_CODE_UP)) {
            heart_offset[1] += heart_speed * levo_portal_my_imports_delta_seconds();
        }
    
        if (levo_portal_my_imports_key_pressed(LEVO_PORTAL_MY_IMPORTS_KEY_CODE_DOWN)) {
            heart_offset[1] -= heart_speed * levo_portal_my_imports_delta_seconds();
        }
    
        drawHeart(heart_offset[0], heart_offset[1]);
    }
    
    if (tick > 200) {
        my_world_string_t my_string;
        my_world_string_set(&my_string, "Happy New Year from C!");
    
        my_world_string_t label_color;
        my_world_string_set(&label_color, "white");

        my_world_string_t link_url;
        my_world_string_set(&link_url, "localhost/go.wasm");

        my_world_string_t link_name;
        my_world_string_set(&link_name, "Go to go.wasm");

        levo_portal_my_imports_label(&my_string, 0.0, -200.0, 64.0, &label_color);
        levo_portal_my_imports_link(&link_url, &link_name, -100.0, -300.0, 32.0);
    }
}

void my_world_setup() {
    my_world_string_t my_string;
    my_world_string_set(&my_string, "setup from guest (C) has been called");

    levo_portal_my_imports_print(&my_string);
}

File: clients/go-client-app/src/my-component.go
// my-component.go
package main

import (
        "math"
        "math/rand"
        world "workspace/my-world"
)

type particle struct {
        x      float32
        y      float32
        speed  float32
        radius float32
        color  string
}

var particles []particle

var tick int = 0
var heartOffset = [2]float32{0.0, 0.0}

func createParticles() {
    canvasSize := world.LevoPortalMyImportsCanvasSize()
    canvasWidth := canvasSize.Width
    tick++

    leftMouseButton := world.LevoPortalMyImportsMouseButtonLeft()

    if tick%10 == 0 {
        if len(particles) < 100 {
            newParticle := particle{
                x:      rand.Float32() * canvasWidth,
                y:      0,
                speed:  500 + rand.Float32()*13,
                radius: 5 + rand.Float32()*5,
                color:  "white",
            }

            cursorPositionOption := world.LevoPortalMyImportsCursorPosition()

            if cursorPositionOption.IsSome() && world.LevoPortalMyImportsMouseButtonPressed(leftMouseButton) {
                cursorPosition := cursorPositionOption.Unwrap()
                newParticle.x = cursorPosition.X
                newParticle.y = cursorPosition.Y
            }

            particles = append(particles, newParticle)
        }
    }
}

func updateParticles() {
        for i := 0; i < len(particles); i++ {
                particle := &particles[i]
                particle.y -= particle.speed * world.LevoPortalMyImportsDeltaSeconds()
        }
}

func killParticles() {
	var canvas_height float32 = world.LevoPortalMyImportsCanvasSize().Height
        for i := 0; i < len(particles); i++ {
                particle := &particles[i]
                if particle.y < -canvas_height {
                        particle.y = 0
                }
        }
}

func drawParticles() {
	var canvas_width float32 = world.LevoPortalMyImportsCanvasSize().Width
	var canvas_height float32 = world.LevoPortalMyImportsCanvasSize().Height
        world.LevoPortalMyImportsFillStyle("royal_purple")
        world.LevoPortalMyImportsFillRect(0, 0, canvas_width, canvas_height)
        for i := 0; i < len(particles); i++ {
                particle := &particles[i]
                world.LevoPortalMyImportsBeginPath()
                world.LevoPortalMyImportsArc(
                        particle.x,
                        particle.y,
                        particle.radius,
                        2*math.Pi,
                        0,
                )
                world.LevoPortalMyImportsClosePath()
                world.LevoPortalMyImportsFillStyle(particle.color)
                world.LevoPortalMyImportsFill()
        }
}

func drawHeart(xOffset float32, yOffset float32) {
    world.LevoPortalMyImportsBeginPath()
    world.LevoPortalMyImportsMoveTo(xOffset, yOffset)
    world.LevoPortalMyImportsCubicBezierTo(xOffset+70, yOffset+70, xOffset+175, yOffset-35, xOffset, yOffset-140)
    world.LevoPortalMyImportsCubicBezierTo(xOffset-175, yOffset-35, xOffset-70, yOffset+70, xOffset, yOffset)
    world.LevoPortalMyImportsClosePath()
    world.LevoPortalMyImportsFillStyle("red")
    world.LevoPortalMyImportsFill()
}

func init() {
        a := HostImpl{}
        world.SetMyWorld(a)
}

type HostImpl struct {
}

func (e HostImpl) Setup() {
        world.LevoPortalMyImportsPrint("setup from guest (Go) has been called")
}

func (e HostImpl) Update() {
    createParticles()
    updateParticles()
    killParticles()
    drawParticles()
    
    if tick > 100 {
        heartSpeed := float32(222.0)
        
	    if world.LevoPortalMyImportsKeyPressed(world.LevoPortalMyImportsKeyCodeLeft()) {
            heartOffset[0] -= heartSpeed * world.LevoPortalMyImportsDeltaSeconds()
        }
        
	    if world.LevoPortalMyImportsKeyPressed(world.LevoPortalMyImportsKeyCodeRight()) {
            heartOffset[0] += heartSpeed * world.LevoPortalMyImportsDeltaSeconds()
        }
        
	    if world.LevoPortalMyImportsKeyPressed(world.LevoPortalMyImportsKeyCodeUp()) {
            heartOffset[1] += heartSpeed * world.LevoPortalMyImportsDeltaSeconds()
        }
        
	    if world.LevoPortalMyImportsKeyPressed(world.LevoPortalMyImportsKeyCodeDown()) {
            heartOffset[1] -= heartSpeed * world.LevoPortalMyImportsDeltaSeconds()
        }
        
        drawHeart(heartOffset[0], heartOffset[1])
    }

    if tick > 200 {
        world.LevoPortalMyImportsLabel("Happy New Year from Go!", 0.0, -200.0, 64.0, "white")
        world.LevoPortalMyImportsLink("../rust.wasm", "Go to rust.wasm", -100.0, -300.0, 32.0)
    }
}

func main() {}
